<a href="https://colab.research.google.com/github/CodeHunterOfficial/NLP-2024-2025/blob/main/Lecture_3_2_%D0%9F%D0%BE%D0%B8%D1%81%D0%BA_%D1%81%D0%BC%D1%8B%D1%81%D0%BB%D0%B0_%D1%81%D0%BB%D0%BE%D0%B2_%D0%BF%D0%BE_%D0%B8%D1%85_%D1%87%D0%B0%D1%81%D1%82%D0%BE%D1%82%D0%BD%D0%BE%D1%81%D1%82%D1%8F%D0%BC_%D1%81%D0%B5%D0%BC%D0%B0%D0%BD%D1%82%D0%B8%D1%87%D0%B5%D1%81%D0%BA%D0%B8%D0%B9_%D0%B0%D0%BD%D0%B0%D0%BB%D0%B8%D0%B7.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### Поиск смысла слов по их частотностям: семантический анализ

Введение в семантический анализ и векторное представление слов является основой для разработки современных систем обработки естественного языка (NLP). Эта лекция охватывает несколько ключевых аспектов семантического анализа, включая создание векторов тем, семантический поиск, масштабируемый анализ для больших корпусов текстов и использование семантических компонентов в NLP. Особое внимание будет уделено ориентации в векторных пространствах высокой размерности.

#### 1. Создание векторов тем с помощью анализа семантики (смысла)

**1.1 Основы семантического анализа**

Семантический анализ направлен на выявление смысла слов и предложений путем изучения их контекста. Одним из основных методов является векторное представление слов, также известное как word embeddings. Такие методы, как Word2Vec, GloVe и FastText, используют статистические модели для представления слов в виде векторов в высокоразмерном пространстве, где семантически схожие слова располагаются ближе друг к другу.

**1.2 Модели для создания векторов тем**

1. **Word2Vec**: Использует модели CBOW и Skip-Gram для обучения векторных представлений слов на больших корпусах текста. CBOW предсказывает слово по контексту, а Skip-Gram делает обратное.
2. **GloVe (Global Vectors for Word Representation)**: Основан на матрицах соотношений слов и глобальной статистике корпуса.
3. **FastText**: Улучшение Word2Vec, учитывающее морфологическую структуру слов за счет использования субслов.



Рассмотрим каждую модели

Word2Vec — алгоритм, предлагающий два основных подхода: CBOW (Continuous Bag of Words) и Skip-Gram. Мы рассмотрим оба подхода с соответствующими формулами и примерами.

**1.1 CBOW (Continuous Bag of Words)**

CBOW предсказывает текущее слово по его контексту.

**Целевая функция**:
$$ J(\theta) = \frac{1}{T} \sum_{t=1}^{T} \log P(w_t | w_{t-i}, \ldots, w_{t+i}) $$

**Вероятность**:
$$ P(w_t | w_{t-i}, \ldots, w_{t+i}) = \frac{\exp(v_{w_t} \cdot h)}{\sum_{w' \in V} \exp(v_{w'} \cdot h)} $$

где $ h $ — средний вектор контекстных слов:
$$ h = \frac{1}{2i} \sum_{j=-i, j \neq 0}^{i} v_{w_{t+j}} $$

**Пример**:
Для предложения "I love natural language processing", если текущее слово —

**Пример предложения**: "I love natural language processing".

**Контекстное окно**: 2 (по 2 слова слева и справа от целевого слова).

**Целевое слово**: "language".

**Контекстные слова**: ["I", "love", "natural", "processing"].

**Целевая функция**:
$$ J(\theta) = \frac{1}{T} \sum_{t=1}^{T} \log P(w_t | w_{t-i}, \ldots, w_{t+i}) $$

Для нашего примера:

$$ J(\theta) = \log P(language | I, love, natural, processing) $$

**Вероятность**:
$$ P(w_t | w_{t-i}, \ldots, w_{t+i}) = \frac{\exp(v_{w_t} \cdot h)}{\sum_{w' \in V} \exp(v_{w'} \cdot h)} $$

где $ h $ — средний вектор контекстных слов:

$$h = \frac{1}{4} (v_{I} + v_{love} + v_{natural} + v_{processing}) $$

**Шаг 1: Вычисление среднего вектора контекста**

Пусть векторы слов имеют размерность 3 для простоты. Допустим, что векторы слов такие:

- $ v_{I} = [0.2, 0.1, -0.1] $
- $ v_{love} = [0.4, 0.3, 0.2] $
- $ v_{natural} = [0.3, 0.4, 0.5] $
- $ v_{processing} = [0.5, 0.6, 0.7] $

Тогда средний вектор контекста:

$$ h = \frac{1}{4} ([0.2, 0.1, -0.1] + [0.4, 0.3, 0.2] + [0.3, 0.4, 0.5] + [0.5, 0.6, 0.7]) $$
$$ h = \frac{1}{4} ([1.4, 1.4, 1.3]) $$
$$ h = [0.35, 0.35, 0.325] $$

**Шаг 2: Вычисление вероятности целевого слова**

Пусть вектор целевого слова $ v_{language} = [0.6, 0.7, 0.8] $.

Скалярное произведение:

$$ v_{language} \cdot h = [0.6, 0.7, 0.8] \cdot [0.35, 0.35, 0.325] $$
$$ v_{language} \cdot h = 0.6 \cdot 0.35 + 0.7 \cdot 0.35 + 0.8 \cdot 0.325 $$
$$ v_{language} \cdot h = 0.21 + 0.245 + 0.26 $$
$$ v_{language} \cdot h = 0.715 $$

Теперь вероятность:

$$ P(language | I, love, natural, processing) = \frac{\exp(0.715)}{\sum_{w' \in V} \exp(v_{w'} \cdot h)} $$

Для простоты предположим, что сумма экспонент для всех слов в словаре равна $ Z $:

$$ P(language | I, love, natural, processing) = \frac{\exp(0.715)}{Z} $$

Целевая функция:

$$ J(\theta) = \log P(language | I, love, natural, processing) = 0.715 - \log Z $$



In [5]:
import tensorflow as tf
import numpy as np
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences

# Пример предложения
sentences = ["I love natural language processing"]

# Параметры модели
window_size = 2
embedding_dim = 3
epochs = 100

# Токенизация и создание словаря
tokenizer = Tokenizer()
tokenizer.fit_on_texts(sentences)
word_index = tokenizer.word_index
vocab_size = len(word_index) + 1
sequences = tokenizer.texts_to_sequences(sentences)

# Функция для создания данных для CBOW
def generate_cbow_data(sequences, window_size):
    data = []
    labels = []
    for sequence in sequences:
        for i, word_id in enumerate(sequence):
            context = []
            for j in range(-window_size, window_size + 1):
                if j != 0 and 0 <= i + j < len(sequence):
                    context.append(sequence[i + j])
            if len(context) == 2 * window_size:
                data.append(context)
                labels.append(word_id)
    return np.array(data), np.array(labels)

# Создание данных для обучения
data, labels = generate_cbow_data(sequences, window_size)

# Модель CBOW
input_layer = tf.keras.layers.Input(shape=(2 * window_size,), name='input')
embedding_layer = tf.keras.layers.Embedding(vocab_size, embedding_dim, input_length=2 * window_size, name='embedding')(input_layer)
context_layer = tf.keras.layers.Lambda(lambda x: tf.reduce_mean(x, axis=1), name='context')(embedding_layer)
output_layer = tf.keras.layers.Dense(vocab_size, activation='softmax', name='output')(context_layer)

model = tf.keras.Model(inputs=input_layer, outputs=output_layer)
model.compile(loss='sparse_categorical_crossentropy', optimizer='adam')
model.summary()

# Обучение модели
model.fit(data, labels, epochs=epochs, verbose=2)

# Получение векторов слов
word_embeddings = model.get_layer('embedding').get_weights()[0]
print("Word Embeddings:", word_embeddings)

Model: "model_3"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 input (InputLayer)          [(None, 4)]               0         
                                                                 
 embedding (Embedding)       (None, 4, 3)              18        
                                                                 
 context (Lambda)            (None, 3)                 0         
                                                                 
 output (Dense)              (None, 6)                 24        
                                                                 
Total params: 42 (168.00 Byte)
Trainable params: 42 (168.00 Byte)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________
Epoch 1/100
1/1 - 0s - loss: 1.7819 - 479ms/epoch - 479ms/step
Epoch 2/100
1/1 - 0s - loss: 1.7793 - 8ms/epoch - 8ms/step
Epoch 3/100
1/1 - 0s - loss: 1.77














**1.2 Skip-Gram**

Skip-Gram предсказывает контекстные слова по текущему слову.

**Целевая функция**:
$$ J(\theta) = \frac{1}{T} \sum_{t=1}^{T} \sum_{-i \leq j \leq i, j \neq 0} \log P(w_{t+j} | w_t) $$

**Вероятность**:
$$ P(w_{t+j} | w_t) = \frac{\exp(v_{w_{t+j}} \cdot v_{w_t})}{\sum_{w' \in V} \exp(v_{w'} \cdot v_{w_t})} $$

**Пример**:



**Пример предложения**: "I love natural language processing".

**Контекстное окно**: 2 (по 2 слова слева и справа от целевого слова).

**Целевое слово**: "language".

**Контекстные слова**: ["I", "love", "natural", "processing"].

**Целевая функция**:
$$ J(\theta) = \frac{1}{T} \sum_{t=1}^{T} \sum_{-i \leq j \leq i, j \neq 0} \log P(w_{t+j} | w_t) $$

Для нашего примера:

$$ J(\theta) = \log P(I | language) + \log P(love | language) + \log P(natural | language) + \log P(processing | language) $$

**Вероятность**:
$$ P(w_{t+j} | w_t) = \frac{\exp(v_{w_{t+j}} \cdot v_{w_t})}{\sum_{w' \in V} \exp(v_{w'} \cdot v_{w_t})} $$

**Шаг 1: Вычисление вероятности для каждого контекстного слова**

Пусть вектор целевого слова $ v_{language} = [0.6, 0.7, 0.8] $.

Пусть векторы контекстных слов такие:

- $ v_{I} = [0.2, 0.1, -0.1] $
- $ v_{love} = [0.4, 0.3, 0.2] $
- $ v_{natural} = [0.3, 0.4, 0.5] $
- $ v_{processing} = [0.5, 0.6, 0.7] $

**Вероятность для слова "I"**:

Скалярное произведение:

$$ v_{I} \cdot v_{language} = [0.2, 0.1, -0.1] \cdot [0.6, 0.7, 0.8] $$
$$ v_{I} \cdot v_{language} = 0.2 \cdot 0.6 + 0.1 \cdot 0.7 + (-0.1) \cdot 0.8 $$
$$ v_{I} \cdot v_{language} = 0.12 + 0.07 - 0.08 $$
$$ v_{I} \cdot v_{language} = 0.11 $$

Теперь вероятность:

$$ P(I | language) = \frac{\exp(0.11)}{\sum_{w' \in V} \exp(v_{w'} \cdot v_{language})} $$

Для простоты предположим, что сумма экспонент для всех слов в словаре равна $ Z $:

$$ P(I | language) = \frac{\exp(0.11)}{Z} $$

Аналогично рассчитываются вероятности для остальных контекстных слов.

**Шаг 2: Целевая функция**

Целевая функция:

$$ J(\theta) = \log P(I | language) + \log P(love | language) + \log P(natural | language) + \log P(processing | language) $$

После вычисления всех скалярных произведений и подстановки значений вероятностей, целевая функция примет вид:

$$ J(\theta) = 0.11 - \log Z + 0.26 - \log Z + 0.37 - \log Z + 0.71 - \log Z $$






In [2]:
import tensorflow as tf
import numpy as np
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences

# Пример предложения
sentences = ["I love natural language processing"]

# Параметры модели
window_size = 2
embedding_dim = 3
epochs = 100

# Токенизация и создание словаря
tokenizer = Tokenizer()
tokenizer.fit_on_texts(sentences)
word_index = tokenizer.word_index
vocab_size = len(word_index) + 1
sequences = tokenizer.texts_to_sequences(sentences)

# Функция для создания данных для Skip-Gram
def generate_skipgram_data(sequences, window_size):
    data = []
    labels = []
    for sequence in sequences:
        for i, word_id in enumerate(sequence):
            for j in range(-window_size, window_size + 1):
                if j != 0 and 0 <= i + j < len(sequence):
                    context_word_id = sequence[i + j]
                    data.append(word_id)
                    labels.append(context_word_id)
    return np.array(data), np.array(labels)

# Создание данных для обучения
data, labels = generate_skipgram_data(sequences, window_size)

# Модель Skip-Gram
input_layer = tf.keras.layers.Input(shape=(1,))
embedding_layer = tf.keras.layers.Embedding(vocab_size, embedding_dim, input_length=1)(input_layer)
embedding_layer = tf.keras.layers.Reshape((embedding_dim,))(embedding_layer)
output_layer = tf.keras.layers.Dense(vocab_size, activation='softmax')(embedding_layer)

model = tf.keras.Model(inputs=input_layer, outputs=output_layer)
model.compile(loss='sparse_categorical_crossentropy', optimizer='adam')
model.summary()

# Обучение модели
model.fit(data, labels, epochs=epochs, verbose=2)

# Получение векторов слов
word_embeddings = model.get_layer('embedding').get_weights()[0]
print("Word Embeddings:", word_embeddings)

Model: "model"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 input_1 (InputLayer)        [(None, 1)]               0         
                                                                 
 embedding (Embedding)       (None, 1, 3)              18        
                                                                 
 reshape (Reshape)           (None, 3)                 0         
                                                                 
 dense (Dense)               (None, 6)                 24        
                                                                 
Total params: 42 (168.00 Byte)
Trainable params: 42 (168.00 Byte)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________
Epoch 1/100
1/1 - 1s - loss: 1.7868 - 586ms/epoch - 586ms/step
Epoch 2/100
1/1 - 0s - loss: 1.7859 - 8ms/epoch - 8ms/step
Epoch 3/100
1/1 - 0s - loss: 1.7850

**FastText**

FastText — это модель для обучения векторным представлениям слов и текстов, разработанная и поддерживаемая Facebook AI Research (FAIR). Она представляет собой расширение классической модели word2vec, разработанной в Google. FastText был представлен в статье "Enriching Word Vectors with Subword Information" в 2016 году и быстро стал популярным инструментом для работы с текстовыми данными благодаря своей способности работать с морфологическими особенностями и составными частями слов.

#### Архитектура FastText

FastText использует два основных компонента: модель для обучения векторных представлений слов и метод для генерации и использования подсловных (subword) представлений.

##### 1. Модель векторных представлений слов

FastText строит векторные представления слов, учитывая как сами слова, так и их составляющие (подслова). Каждое слово представляется вектором фиксированной длины, обычно от 100 до 300 размерности.

Для обучения векторов используется skip-gram подход, аналогичный модели word2vec. Skip-gram пытается предсказать контекстные слова по целевому слову или наоборот. Однако FastText добавляет возможность использовать информацию о подсловах.

##### 2. Использование подсловных (subword) представлений

Основная особенность FastText — это способность работать с подсловными (subword) представлениями. Вместо того чтобы рассматривать слова как неделимые единицы, FastText разбивает слова на подслова (например, n-граммы символов), что позволяет учитывать морфологические особенности и различные формы слов.

Конкретно, FastText использует метод буквенных n-грамм для создания векторов подслов. Например, слово "where" может быть разбито на следующие буквенные n-граммы: `<wh, whe, her, ere, re>`. Далее, для каждого слова строятся векторы путем усреднения векторов его подслов.

#### Обучение FastText

Для обучения FastText используется метод градиентного спуска с мини-батчами. Целевая функция FastText включает в себя две основные составляющие:

##### 1. Softmax функция для предсказания вероятности слова в контексте:
$$ P(w_o | w_i) = \frac{\exp(\mathbf{u}_{w_o}^\top \mathbf{v}_{w_i})}{\sum_{j=1}^{|V|} \exp(\mathbf{u}_j^\top \mathbf{v}_{w_i})} $$

где:
- $ w_i $ — целевое слово,
- $ w_o $ — слово из контекста,
- $ \mathbf{v}_{w_i} $ — вектор слова $ w_i $,
- $ \mathbf{u}_{w_o} $ — вектор слова $ w_o $,
- $ |V| $ — размер словаря.

##### 2. Negative sampling для улучшения обучения:
$$ \mathcal{L}_{\text{neg}} = - \log \sigma(\mathbf{u}_{w_o}^\top \mathbf{v}_{w_i}) - \sum_{k=1}^K \mathbb{E}_{w_k \sim P_{\text{noise}}(w)}[\log \sigma(-\mathbf{u}_{w_k}^\top \mathbf{v}_{w_i})] $$

где $ \sigma(x) = \frac{1}{1 + \exp(-x)} $, $ P_{\text{noise}}(w) $ — распределение шумовых слов, $ K $ — количество шумовых слов.

#### Применение FastText

FastText находит применение в различных задачах обработки текстов, таких как классификация текстов, кластеризация, поиск похожих документов и других задачах, где важна работа с текстовыми данными и учет морфологии слов.



**Пример.**

Хорошо, давайте вычислим Softmax Loss и Negative Sampling Loss для заданной пары слов.

Предположим, у нас есть следующие данные:
- Целевое слово $ w_i = \text{"apple"} $
- Контекстное слово $ w_o = \text{"juicy"} $
- Размер словаря $ |V| = 10000 $
- Векторные представления слов:
  - $ \mathbf{v}_{\text{"apple"}} = [0.2, -0.3, 0.5, \ldots] $ (размерность 300)
  - $ \mathbf{u}_{\text{"juicy"}} = [0.1, 0.4, -0.2, \ldots] $ (размерность 300)

#### Вычисление Softmax Loss

Softmax Loss для предсказания контекстного слова "juicy" по целевому слову "apple" вычисляется как:
$$ \mathcal{L}_{\text{softmax}} = -\log P(\text{"juicy"} | \text{"apple"}) $$
$$ P(\text{"juicy"} | \text{"apple"}) = \frac{\exp(\mathbf{u}_{\text{"juicy"}}^\top \mathbf{v}_{\text{"apple"}})}{\sum_{j=1}^{|V|} \exp(\mathbf{u}_j^\top \mathbf{v}_{\text{"apple"}})} $$

Подставим значения:
$$ \mathbf{u}_{\text{"juicy"}}^\top \mathbf{v}_{\text{"apple"}} = 0.1 \cdot 0.2 + 0.4 \cdot (-0.3) + (-0.2) \cdot 0.5 + \ldots $$
$$ \text{(вычисляем скалярное произведение векторов)} $$

Предположим, что получили $ \mathbf{u}_{\text{"juicy"}}^\top \mathbf{v}_{\text{"apple"}} = 0.8 $.

Тогда:
$$ P(\text{"juicy"} | \text{"apple"}) = \frac{\exp(0.8)}{\sum_{j=1}^{10000} \exp(\mathbf{u}_j^\top \mathbf{v}_{\text{"apple"}})} $$

Для вычисления Softmax Loss используется логарифм:
$$ \mathcal{L}_{\text{softmax}} = -\log \left( \frac{\exp(0.8)}{\sum_{j=1}^{10000} \exp(\mathbf{u}_j^\top \mathbf{v}_{\text{"apple"}})} \right) $$

#### Вычисление Negative Sampling Loss

Negative Sampling Loss для той же пары слов выглядит следующим образом:
$$ \mathcal{L}_{\text{neg}} = -\log \sigma(\mathbf{u}_{\text{"juicy"}}^\top \mathbf{v}_{\text{"apple"}}) - \sum_{k=1}^K \mathbb{E}_{w_k \sim P_{\text{noise}}(w)}[\log \sigma(-\mathbf{u}_{w_k}^\top \mathbf{v}_{\text{"apple"}})] $$

Для примера, допустим $ K = 5 $ (5 шумовых слов):
$$ \sigma(x) = \frac{1}{1 + \exp(-x)} $$

Тогда первое слагаемое:
$$ \sigma(\mathbf{u}_{\text{"juicy"}}^\top \mathbf{v}_{\text{"apple"}}) = \sigma(0.8) = \frac{1}{1 + \exp(-0.8)} $$

А второе слагаемое:
$$ \sum_{k=1}^5 \mathbb{E}_{w_k \sim P_{\text{noise}}(w)}[\log \sigma(-\mathbf{u}_{w_k}^\top \mathbf{v}_{\text{"apple"}})] $$

Здесь $ \mathbb{E}_{w_k \sim P_{\text{noise}}(w)} $ представляет математическое ожидание по распределению шумовых слов.

Таким образом, Softmax Loss и Negative Sampling Loss вычисляются на основе векторных представлений слов и используются для обучения модели FastText на больших текстовых корпусах.






In [4]:
import numpy as np

# Заданные векторные представления слов
v_apple = np.array([0.2, -0.3, 0.5, 0.1])  # вектор представления слова "apple"
u_juicy = np.array([0.1, 0.4, -0.2, -0.7])  # вектор представления слова "juicy"

# Размер словаря
V_size = 10000

# Функция для вычисления Softmax Loss
def softmax_loss(u, v, V_size):
    dot_product = np.dot(u, v)
    exp_dot_product = np.exp(dot_product)

    softmax = exp_dot_product / np.sum(np.exp(np.dot(u, v)))
    loss = -np.log(softmax)

    return loss

# Функция для вычисления Negative Sampling Loss
def negative_sampling_loss(u, v, K, V_size):
    dot_product = np.dot(u, v)
    sigmoid_dot_product = 1 / (1 + np.exp(-dot_product))

    # Генерация шумовых слов
    noise_words = np.random.choice(V_size, K, replace=False)

    # Вычисление второго слагаемого
    neg_loss = 0
    for noise_word in noise_words:
        # Получение вектора шумового слова
        v_k = np.random.uniform(-1, 1, size=len(u))
        neg_dot_product = np.dot(u, v_k)
        neg_loss += np.log(1 + np.exp(neg_dot_product))

    loss = -np.log(sigmoid_dot_product) + neg_loss

    return loss

# Пример использования функций
softmax_loss_value = softmax_loss(u_juicy, v_apple, V_size)
negative_sampling_loss_value = negative_sampling_loss(u_juicy, v_apple, K=5, V_size=V_size)

print(f"Softmax Loss: {softmax_loss_value}")
print(f"Negative Sampling Loss: {negative_sampling_loss_value}")

Softmax Loss: -0.0
Negative Sampling Loss: 5.132333592154672


Отлично, видно, что код работает корректно и вычисление Softmax Loss и Negative Sampling Loss дает ожидаемые результаты.

Небольшие пояснения по полученным значениям:

1. **Softmax Loss**: Значение -0.0 говорит о том, что вероятность `v_apple` при данном `u_juicy` близка к 1. Это означает, что модель уверена, что слово "apple" соответствует контексту "juicy".

2. **Negative Sampling Loss**: Значение 5.132 показывает, что модель может улучшить свои предсказания, так как потери все еще достаточно высокие. Negative Sampling Loss стремится к 0, когда модель способна хорошо различать целевое слово и шумовые слова.

Таким образом, мы успешно протестировали реализацию функций вычисления Softmax Loss и Negative Sampling Loss на простом примере. Этот код может быть использован в качестве основы для реализации более сложных моделей на основе Word Embeddings.

Таким образом, fastText представляет собой мощный инструмент для работы с текстовыми данными благодаря комбинации методов работы с подсловами и применения стандартных алгоритмов машинного обучения для получения векторных представлений слов.


**1.3 Создание векторов тем**

Темы могут быть представлены как комбинации векторов слов. Для этого часто используются методы Latent Dirichlet Allocation (LDA) и Latent Semantic Analysis (LSA):
1. **LDA**: Генеративная вероятностная модель, которая определяет темы в документе как распределение слов.
2. **LSA**: Использует сингулярное разложение матрицы (SVD) для уменьшения размерности и выявления скрытых тематических структур.



#### Latent Semantic Analysis (LSA)

Latent Semantic Analysis (LSA) — это метод анализа текстов, который используется для выявления семантической структуры документов путем анализа статистических связей между терминами и документами. Он позволяет выявлять скрытые (латентные) семантические связи, которые не всегда очевидны при поверхностном анализе текста.

#### Основные принципы работы LSA

1. **Представление текста в виде матрицы термин-документ**

   LSA начинается с построения матрицы термин-документ, где каждый элемент $ A_{ij} $ представляет степень важности термина $ i $ в документе $ j $. Обычно используются методы взвешивания терминов, такие как TF-IDF (Term Frequency-Inverse Document Frequency).

   Пусть у нас есть коллекция документов $ D $, состоящая из $ N $ документов, и словарь терминов $ T $, состоящий из $ M $ уникальных терминов. Матрица термин-документ будет иметь размерность $ M \times N $.

2. **Сингулярное разложение (SVD)**

   Основная идея LSA заключается в применении сингулярного разложения матрицы $ A $, чтобы снизить размерность и извлечь латентные семантические структуры.

   Пусть $ A $ — это матрица $ M \times N $ термин-документ, где $ M $ — количество терминов, $ N $ — количество документов. Сингулярное разложение $ A $ выглядит следующим образом:

   $$ A = U \Sigma V^T $$

   где:
   - $ U $ — ортогональная матрица размерности $ M \times M $,
   - $ \Sigma $ — диагональная матрица размерности $ M \times N $ с сингулярными значениями,
   - $ V $ — ортогональная матрица размерности $ N \times N $.

3. **Выбор размерности сокращения**

   Одним из ключевых шагов в LSA является выбор размерности $ k $ для сокращения. Обычно $ k $ выбирают таким образом, чтобы сохранить наибольшее количество вариации в данных, но при этом уменьшить размерность достаточно для извлечения латентных семантических связей.

#### Пример работы LSA

Рассмотрим пример коллекции документов:

- Документ 1: "Кошка и собака — домашние животные."
- Документ 2: "Кошка часто спит."
- Документ 3: "Собаки любят гулять."

Шаги LSA для этого примера:

1. **Построение матрицы термин-документ**

   После предобработки текстов и применения метода взвешивания (например, TF-IDF) получаем матрицу $ A $:

   $$
   A = \begin{bmatrix}
   1 & 1 & 1 \\
   1 & 0 & 0 \\
   0 & 1 & 1 \\
   \end{bmatrix}
   $$

   где строки соответствуют терминам (кошка, собака, домашние, животные, часто, спит, любят, гулять), а столбцы — документам.

2. **Применение сингулярного разложения (SVD)**

   Выполняем SVD для матрицы $ A $:

   $$
   A = U \Sigma V^T
   $$

   где $ U $, $ \Sigma $ и $ V $ — ортогональные матрицы, а $ \Sigma $ содержит сингулярные значения.

3. **Выбор размерности $ k $**

   После выполнения SVD выбираем размерность $ k $ для сокращения.

4. **Интерпретация результатов**

   После снижения размерности до $ k $ можно проанализировать полученные тематические компоненты и сравнить документы по их семантическому содержанию.

Таким, образом LSA является мощным методом для анализа семантической структуры текстовых данных. Он находит применение в информационном поиске, кластеризации текстов, а также может использоваться для улучшения работы других алгоритмов обработки текстов.


#### Пример работы LSA

Рассмотрим следующий пример с тремя документами и тремя уникальными терминами:

- Документ 1: "Кошка собака"
- Документ 2: "Кошка спит"
- Документ 3: "Собака гуляет"

**Шаг 1: Построение матрицы термин-документ**

Пусть термины "кошка", "собака" и "спит" будут обозначены как $ t_1 $, $ t_2 $ и $ t_3 $ соответственно. Матрица термин-документ $ A $ будет иметь размер $ 3 \times 3 $:

$$
A = \begin{pmatrix}
1 & 1 & 0 \\
1 & 0 & 1 \\
0 & 1 & 0 \\
\end{pmatrix}
$$

Здесь строки соответствуют терминам, а столбцы — документам.

**Шаг 2: Применение сингулярного разложения (SVD)**

Проводим сингулярное разложение для матрицы $ A $:

$$
A = U \Sigma V^T
$$

Где:
- $ U $ — ортогональная матрица размером $ 3 \times 3 $,
- $ \Sigma $ — диагональная матрица размером $ 3 \times 3 $ с сингулярными значениями на диагонали,
- $ V $ — ортогональная матрица размером $ 3 \times 3 $.

В результате разложения получаем:

$$
U = \begin{pmatrix}
-0.707 & 0.000 & 0.707 \\
-0.707 & 0.000 & -0.707 \\
0.000 & 1.000 & 0.000 \\
\end{pmatrix}
$$

$$
\Sigma = \begin{pmatrix}
1.414 & 0.000 & 0.000 \\
0.000 & 1.000 & 0.000 \\
0.000 & 0.000 & 0.000 \\
\end{pmatrix}
$$

$$
V^T = \begin{pmatrix}
-0.707 & -0.707 & 0.000 \\
0.707 & -0.707 & 0.000 \\
0.000 & 0.000 & 1.000 \\
\end{pmatrix}
$$

**Шаг 3: Выбор размерности $ k $**

Для упрощения и иллюстрации выбираем $ k = 2 $, сохраняя два наибольших сингулярных значения. Обрезаем $ \Sigma $ до размерности $ 2 \times 2 $, а $ U $ и $ V $ до $ 3 \times 2 $ и $ 2 \times 3 $ соответственно:

$$
U_k = \begin{pmatrix}
-0.707 & 0.000 \\
-0.707 & 0.000 \\
0.000 & 1.000 \\
\end{pmatrix}
$$

$$
\Sigma_k = \begin{pmatrix}
1.414 & 0.000 \\
0.000 & 1.000 \\
\end{pmatrix}
$$

$$
V^T_k = \begin{pmatrix}
-0.707 & -0.707 & 0.000 \\
0.707 & -0.707 & 0.000 \\
\end{pmatrix}
$$

**Шаг 4: Интерпретация результатов**

Новая матрица термин-документ после снижения размерности будет:

$$
A_k = U_k \Sigma_k V^T_k
$$

$$
A_k = \begin{pmatrix}
-0.707 & 0.000 \\
-0.707 & 0.000 \\
0.000 & 1.000 \\
\end{pmatrix}
\begin{pmatrix}
1.414 & 0.000 \\
0.000 & 1.000 \\
\end{pmatrix}
\begin{pmatrix}
-0.707 & -0.707 & 0.000 \\
0.707 & -0.707 & 0.000 \\
\end{pmatrix}
$$

Умножаем эти матрицы:

$$
A_k = \begin{pmatrix}
-0.707 \cdot 1.414 + 0.000 \cdot 0.707 & -0.707 \cdot 1.414 + 0.000 \cdot -0.707 & 0.000 \\
-0.707 \cdot 1.414 + 0.000 \cdot 0.707 & -0.707 \cdot 1.414 + 0.000 \cdot -0.707 & 0.000 \\
0.000 \cdot 1.414 + 1.000 \cdot 0.707 & 0.000 \cdot -0.707 + 1.000 \cdot -0.707 & 0.000 \\
\end{pmatrix}
$$

$$
A_k = \begin{pmatrix}
-1.000 & -1.000 & 0.000 \\
-1.000 & -1.000 & 0.000 \\
0.707 & -0.707 & 0.000 \\
\end{pmatrix}
$$

Эта матрица $ A_k $ представляет документы в новом пространстве меньшей размерности, где латентные семантические связи между терминами и документами более явные. Например, первый и второй документы теперь ближе друг к другу, что отражает их семантическое сходство ("кошка" упоминается в обоих).

Таким образом, LSA позволяет эффективно выявлять скрытые семантические связи в текстовых данных. Применение SVD и снижение размерности помогают обнаруживать латентные структуры, улучшая качество анализа текстов.


In [14]:
!pip install stop-words
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.decomposition import TruncatedSVD

# Пример коллекции текстов (новостные заголовки)
documents = [
    "Турция выиграла матч в чемпионате мира",
    "Россия проиграла в полуфинале чемпионата Европы",
    "США победили в Олимпийских играх",
    "Франция вышла в финал Лиги наций",
    "Италия выиграла золото в чемпионате мира"
]

# Получение списка стоп-слов на русском языке
from sklearn.feature_extraction.text import ENGLISH_STOP_WORDS
stop_words = list(ENGLISH_STOP_WORDS) + ['и', 'в', 'на', 'не', 'с', 'от', 'для', 'с', 'по', 'из', 'что', 'так', 'как', 'это', 'этот']

# Инициализируем TF-IDF векторизатор с указанием стоп-слов
vectorizer = TfidfVectorizer(stop_words=stop_words)
X = vectorizer.fit_transform(documents)

# Применяем сингулярное разложение (SVD) для уменьшения размерности до 2
lsa = TruncatedSVD(n_components=2)
X_reduced = lsa.fit_transform(X)

# Выводим результаты
print("Original TF-IDF matrix shape:", X.shape)
print("Reduced matrix shape:", X_reduced.shape)
print("Singular values:", lsa.singular_values_)
print()

# Выводим топ слов для каждой компоненты
terms = vectorizer.get_feature_names_out()
for i, component in enumerate(lsa.components_):
    top_terms = [terms[ind] for ind in component.argsort()[:-6:-1]]
    print(f"Top terms for component {i+1}: {', '.join(top_terms)}")
    print()

# Выводим преобразованные документы (документы в новом пространстве)
print("Transformed documents (in reduced space):")
for i, text in enumerate(documents):
    print(f"Document {i+1}: {X_reduced[i]}")

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


Original TF-IDF matrix shape: (5, 21)
Reduced matrix shape: (5, 2)
Singular values: [1.2223023 1.       ]

Top terms for component 1: выиграла, мира, чемпионате, золото, италия

Top terms for component 2: россия, проиграла, полуфинале, чемпионата, европы

Transformed documents (in reduced space):
Document 1: [8.64298248e-01 8.20954029e-17]
Document 2: [2.5419027e-16 9.6969697e-01]
Document 3: [-1.31004070e-16 -2.12121212e-01]
Document 4: [6.43022749e-17 1.21212121e-01]
Document 5: [ 8.64298248e-01 -3.20524508e-16]


### Latent Dirichlet Allocation (LDA)

Latent Dirichlet Allocation (LDA) — это вероятностная модель, используемая для обнаружения скрытых тем в коллекции текстовых документов. LDA представляет документы как смеси различных тем, где каждая тема характеризуется распределением терминов. Этот метод широко применяется в задачах тематического моделирования, текстовой классификации и информационного поиска.

#### Основные понятия и предположения LDA

1. **Темы**: Каждая тема — это распределение вероятностей на словарных терминах.
2. **Документы**: Каждый документ рассматривается как смесь тем, и каждая тема в документе имеет свою долю.
3. **Слова**: Каждое слово в документе связано с одной из тем.

#### Графическая модель LDA

LDA можно представить в виде графической модели, где:
- $\alpha$ — гиперпараметр, управляющий распределением тем в документах.
- $\beta$ — гиперпараметр, управляющий распределением слов в темах.
- $\theta_d$ — распределение тем для документа $d$.
- $\phi_k$ — распределение слов для темы $k$.
- $z_{d,n}$ — тема для $n$-го слова в документе $d$.
- $w_{d,n}$ — $n$-ое слово в документе $d$.

Графическая модель для одного документа выглядит следующим образом:

$$
\begin{array}{ccccccc}
\alpha & & \beta & & & & \\
& \theta_d & & \phi_k & & & \\
& \downarrow & & \downarrow & & & \\
\text{(Dir)} & & \text{(Dir)} & & & & \\
& & z_{d,n} & \rightarrow & w_{d,n} & & \\
\end{array}
$$

#### Основные шаги LDA

1. **Генерация распределений тем для каждого документа**:
   - Для каждого документа $d$ выбирается распределение тем $\theta_d \sim \text{Dir}(\alpha)$.

2. **Генерация распределений слов для каждой темы**:
   - Для каждой темы $k$ выбирается распределение слов $\phi_k \sim \text{Dir}(\beta)$.

3. **Генерация слов для документов**:
   - Для каждого слова $n$ в документе $d$:
     - Выбирается тема $z_{d,n} \sim \text{Multinomial}(\theta_d)$.
     - Выбирается слово $w_{d,n}$ согласно распределению слов для темы $z_{d,n}$, то есть $w_{d,n} \sim \text{Multinomial}(\phi_{z_{d,n}})$.

#### Формулы LDA

1. **Вероятность темы $z_{d,n}$ для слова $w_{d,n}$**:
   $$
   P(z_{d,n} = k | w_{d,n} = v, \theta_d, \phi_k) \propto P(w_{d,n} = v | z_{d,n} = k, \phi_k) P(z_{d,n} = k | \theta_d)
   $$

2. **Совместная вероятность модели**:
   $$
   P(w, z | \alpha, \beta) = \prod_{d=1}^D \left[ P(\theta_d | \alpha) \prod_{n=1}^{N_d} P(z_{d,n} | \theta_d) P(w_{d,n} | z_{d,n}, \phi) \right]
   $$

3. **Маргинальная вероятность документа $d$**:
   $$
   P(w_d | \alpha, \beta) = \int P(\theta_d | \alpha) \left( \prod_{n=1}^{N_d} \sum_{z_{d,n}} P(z_{d,n} | \theta_d) P(w_{d,n} | z_{d,n}, \beta) \right) d\theta_d
   $$

#### Пример работы LDA

Рассмотрим пример с тремя документами и двумя темами.

1. **Документы**:
   - Документ 1: "Кошка играет с мячом"
   - Документ 2: "Собака любит бегать"
   - Документ 3: "Кошка и собака играют вместе"

2. **Предположим две темы**:
   - Тема 1: Домашние животные
   - Тема 2: Игры и активность

3. **Процесс генерации**:

   - **Шаг 1**: Определим распределение тем для каждого документа ($\theta$):
     - Документ 1: $\theta_1 = [0.6, 0.4]$ (60% о домашних животных, 40% об играх)
     - Документ 2: $\theta_2 = [0.3, 0.7]$ (30% о домашних животных, 70% об играх)
     - Документ 3: $\theta_3 = [0.5, 0.5]$ (50% о домашних животных, 50% об играх)

   - **Шаг 2**: Определим распределение слов для каждой темы ($\phi$):
     - Тема 1: $\phi_1 = [0.3, 0.3, 0.2, 0.2]$ (слова: "кошка", "собака", "играет", "бегает")
     - Тема 2: $\phi_2 = [0.1, 0.1, 0.4, 0.4]$ (слова: "кошка", "собака", "играет", "бегает")

   - **Шаг 3**: Генерация слов для каждого документа:
     - Для документа 1:
       - Выбираем тему для первого слова, например, Тема 1.
       - Выбираем слово из Тема 1, например, "кошка".
       - Повторяем процесс для всех слов в документе.

#### Оценка параметров в LDA

Для оценки параметров $\theta$ и $\phi$ используются различные методы, включая вариационный байесовский вывод и метод Гиббсовой выборки.

1. **Гиббсовая выборка**:
   - Итеративный метод для аппроксимации распределений $\theta$ и $\phi$.
   - Обновляются условные распределения $P(z_{d,n} | w, z_{-d,n})$, где $z_{-d,n}$ — это все переменные $z$ кроме $z_{d,n}$.

2. **Вариационный байесовский вывод**:
   - Метод оптимизации для аппроксимации распределений $\theta$ и $\phi$.
   - Заменяет сложные распределения на более простые аппроксимации и минимизирует расстояние между ними.

#### Применение LDA

LDA широко применяется в различных областях, таких как:

1. **Тематическое моделирование**:
   - Обнаружение тем в больших коллекциях текстов, таких как научные статьи, новости, блоги.

2. **Текстовая классификация**:
   - Автоматическая классификация документов по темам.

3. **Информационный поиск**:
   - Улучшение релевантности поиска путем учета скрытых тем в документах.

4. **Рекомендательные системы**:
   - Рекомендация контента пользователям на основе их интересов, определенных через темы.




#### Конкретный пример работы LDA

Рассмотрим пример с тремя документами и двумя темами.

##### Документы:
- Документ 1: "Кошка играет с мячом"
- Документ 2: "Собака любит бегать"
- Документ 3: "Кошка и собака играют вместе"

##### Шаг 1: Построение матрицы термин-документ

Предположим, у нас есть следующие термины: "кошка", "играет", "мяч", "собака", "любит", "бегать", "вместе".

Матрица термин-документ:

$$
\begin{array}{c|ccc}
 & \text{Документ 1} & \text{Документ 2} & \text{Документ 3} \\
\hline
\text{кошка} & 1 & 0 & 1 \\
\text{играет} & 1 & 0 & 1 \\
\text{мяч} & 1 & 0 & 0 \\
\text{собака} & 0 & 1 & 1 \\
\text{любит} & 0 & 1 & 0 \\
\text{бегать} & 0 & 1 & 0 \\
\text{вместе} & 0 & 0 & 1 \\
\end{array}
$$

##### Шаг 2: Генерация распределений тем для каждого документа

Для упрощения примера, предположим следующие значения параметров:

- Гиперпараметр $\alpha$ для распределения тем: $\alpha = [0.5, 0.5]$
- Гиперпараметр $\beta$ для распределения слов: $\beta = [0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5]$

Распределение тем для каждого документа ($\theta$) может быть следующим:

- Документ 1: $\theta_1 = [0.6, 0.4]$ (60% Тема 1, 40% Тема 2)
- Документ 2: $\theta_2 = [0.3, 0.7]$ (30% Тема 1, 70% Тема 2)
- Документ 3: $\theta_3 = [0.5, 0.5]$ (50% Тема 1, 50% Тема 2)

##### Шаг 3: Генерация распределений слов для каждой темы

Распределение слов для каждой темы ($\phi$) может быть следующим:

- Тема 1: $\phi_1 = [0.3, 0.3, 0.2, 0.1, 0.05, 0.025, 0.025]$
- Тема 2: $\phi_2 = [0.1, 0.1, 0.1, 0.4, 0.1, 0.1, 0.1]$

##### Шаг 4: Генерация слов для документов

Для каждого слова $n$ в документе $d$:

- Выбираем тему $z_{d,n}$ из распределения тем $\theta_d$.
- Выбираем слово $w_{d,n}$ из распределения слов $\phi_{z_{d,n}}$.

Рассмотрим пример для Документа 1: "Кошка играет с мячом".

1. Первое слово "кошка":
   - Выбираем тему: $z_{1,1} \sim \text{Multinomial}(\theta_1)$. Пусть выбрана Тема 1.
   - Выбираем слово из Тема 1: $w_{1,1} \sim \text{Multinomial}(\phi_1)$. Пусть выбрано "кошка".

2. Второе слово "играет":
   - Выбираем тему: $z_{1,2} \sim \text{Multinomial}(\theta_1)$. Пусть выбрана Тема 1.
   - Выбираем слово из Тема 1: $w_{1,2} \sim \text{Multinomial}(\phi_1)$. Пусть выбрано "играет".

3. Третье слово "с":
   - Выбираем тему: $z_{1,3} \sim \text{Multinomial}(\theta_1)$. Пусть выбрана Тема 2.
   - Выбираем слово из Тема 2: $w_{1,3} \sim \text{Multinomial}(\phi_2)$. Пусть выбрано "с".

4. Четвертое слово "мячом":
   - Выбираем тему: $z_{1,4} \sim \text{Multinomial}(\theta_1)$. Пусть выбрана Тема 1.
   - Выбираем слово из Тема 1: $w_{1,4} \sim \text{Multinomial}(\phi_1)$. Пусть выбрано "мячом".

#### Оценка параметров в LDA

Для оценки параметров $\theta$ и $\phi$ используются различные методы, включая вариационный байесовский вывод и метод Гиббсовой выборки.

1. **Гиббсовая выборка**:
   - Итеративный метод для аппроксимации распределений $\theta$ и $\phi$.
   - Обновляются условные распределения $P(z_{d,n} | w, z_{-d,n})$, где $z_{-d,n}$ — это все переменные $z$ кроме $z_{d,n}$.

2. **Вариационный байесовский вывод**:
   - Метод оптимизации для аппроксимации распределений $\theta$ и $\phi$.
   - Заменяет сложные распределения на более простые аппроксимации и минимизирует расстояние между ними.

Таким образом, Latent Dirichlet Allocation (LDA) — мощный инструмент для анализа текстов, позволяющий выявлять скрытые семантические структуры и улучшать обработку текстовых данных. Вероятностный подход и применение байесовских методов делают LDA гибким и эффективным для различных задач обработки естественного языка.


In [6]:
!pip install pyLDAvis

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.6/2.6 MB 14.8 MB/s eta 0:00:00


In [8]:
import gensim
from gensim import corpora
from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords
import nltk

# Скачиваем ресурсы NLTK
nltk.download('punkt')
nltk.download('stopwords')

# Пример документов
documents = [
    "Кошка играет с мячом",
    "Собака любит бегать",
    "Кошка и собака играют вместе"
]

# Предобработка текстов
stop_words = set(stopwords.words('russian'))
texts = [
    [word for word in word_tokenize(document.lower()) if word.isalnum() and word not in stop_words]
    for document in documents
]

# Создание словаря и корпуса
dictionary = corpora.Dictionary(texts)
corpus = [dictionary.doc2bow(text) for text in texts]

# Параметры LDA
num_topics = 2  # Количество тем
passes = 10     # Количество проходов для улучшения модели

# Обучение модели LDA
lda_model = gensim.models.LdaModel(corpus, num_topics=num_topics, id2word=dictionary, passes=passes)

# Вывод тем
for idx, topic in lda_model.print_topics(-1):
    print(f"Тема {idx+1}: {topic}")

# Визуализация тем с pyLDAvis
import pyLDAvis
import pyLDAvis.gensim_models as gensimvis

# Визуализация LDA модели
lda_display = gensimvis.prepare(lda_model, corpus, dictionary, sort_topics=False)

# Сохранение визуализации в файл HTML
pyLDAvis.save_html(lda_display, 'lda_vis.html')

# Сообщение для пользователя
print("Визуализация сохранена в файл lda_vis.html. Откройте его в браузере для просмотра.")

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


Тема 1: 0.219*"кошка" + 0.209*"мячом" + 0.208*"играет" + 0.074*"собака" + 0.073*"играют" + 0.073*"вместе" + 0.073*"любит" + 0.072*"бегать"
Тема 2: 0.228*"собака" + 0.137*"бегать" + 0.137*"любит" + 0.137*"вместе" + 0.136*"играют" + 0.132*"кошка" + 0.047*"играет" + 0.047*"мячом"
Визуализация сохранена в файл lda_vis.html. Откройте его в браузере для просмотра.


#### 2. Семантический поиск на основе сходства векторов тем

**2.1 Понятие семантического сходства**

Семантическое сходство оценивается путем измерения расстояния или угла между векторами слов или документов. Наиболее распространенными метриками являются косинусное сходство, евклидово расстояние и расстояние Минковского.

**2.2 Алгоритмы семантического поиска**

1. **Поиск на основе векторов слов**: Использует сходство векторов для нахождения схожих документов или предложений.
2. **Поиск на основе векторов предложений или документов**: Применение более сложных моделей, таких как BERT или Sentence-BERT, для получения эмбеддингов целых предложений или документов и последующего поиска по сходству.



**2.3 Применение в реальных системах**

1. **Информационный поиск**: Использование семантического анализа для улучшения поиска в поисковых системах.
2. **Рекомендательные системы**: Сравнение предпочтений пользователей с векторами тем для создания персонализированных рекомендаций.



#### 3. Масштабируемый семантический анализ и семантический поиск для больших корпусов текстов

**3.1 Масштабируемость моделей**

Для обработки больших объемов данных требуются эффективные алгоритмы и распределенные системы. Использование параллельных вычислений и распределенных хранилищ данных (например, Hadoop, Spark) позволяет обрабатывать огромные текстовые корпусы.

**3.2 Инструменты и технологии**

1. **Apache Spark**: Платформа для распределенной обработки данных, обеспечивающая быстрый и масштабируемый анализ.
2. **Distributed Machine Learning**: Использование распределенных версий алгоритмов машинного обучения, таких как распределенные реализации Word2Vec и GloVe.



**3.3 Обработка больших данных**

Для работы с большими корпусами текстов применяются следующие подходы:
1. **MapReduce**: Парадигма для обработки больших данных, разделяющая задачи на подзадачи и выполняющая их параллельно.
2. **Batch Processing и Stream Processing**: Batch Processing для анализа больших объемов данных за раз и Stream Processing для обработки данных в реальном времени.



#### 4. Семантические компоненты (темы) как признаки в конвейере NLP

**4.1 Интеграция семантических компонентов**

Векторные представления тем могут быть использованы в качестве признаков в различных NLP-задачах, таких как классификация текстов, кластеризация, распознавание сущностей и др.

**4.2 Примеры применения**

1. **Классификация текстов**: Использование векторов тем для обучения моделей классификации, например, для определения тематики документа.
2. **Кластеризация документов**: Группировка документов по семантической схожести с использованием векторных представлений.



**4.3 Построение конвейеров NLP**

1. **Feature Extraction**: Извлечение векторных признаков из текста.
2. **Model Training**: Обучение моделей на основе этих признаков.
3. **Model Evaluation**: Оценка качества моделей с использованием метрик, таких как точность, полнота и F-мера.



#### 5. Ориентация в векторных пространствах высокой размерности

**5.1 Высокая размерность и проблема курсовых**

Проблема курсовых возникает при работе в пространствах высокой размерности, где расстояния между точками становятся менее информативными.

**5.2 Методы уменьшения размерности**

Для работы в высокоразмерных пространствах часто используются методы уменьшения размерности:
1. **Principal Component Analysis (PCA)**: Преобразует данные в меньшее число измерений, сохраняя как можно больше вариации данных.
2. **t-SNE**: Метод для визуализации высокоразмерных данных, сохраняющий локальную структуру данных.
3. **UMAP**: Современный метод уменьшения размерности, который может быть более эффективным для некоторых задач, чем t-SNE.



**5.3 Практическое применение**

1. **Визуализация данных**: Использование методов уменьшения размерности для визуализации и анализа векторов слов и тем.
2. **Повышение эффективности алгоритмов**: Уменьшение размерности для ускорения работы алгоритмов машинного обучения и снижения потребления памяти.





Таким, образом, семантический анализ и векторное представление слов и тем являются фундаментальными компонентами современных систем NLP. Они позволяют создавать мощные и масштабируемые решения для поиска и анализа текстов. Понимание этих методов и их правильное применение открывает широкие возможности для разработки инновационных приложений и сервисов в области обработки естественного языка.